src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
  * 02111-1307, USA.
  *)
 
-(** Module [Lwt]: cooperative light-weight threads. *)
 
-(** This module defines {e cooperative light-weight threads} with
-    their primitives. A {e light-weight thread} represent a
-    computation that may be not terminated, for example because it is
-    waiting for some event to happen.
 
-    Lwt threads are cooperative in the sense that switching to another
-    thread is always explicit (with {!wakeup} or {!wakeup_exn}). When a
-    thread is running, it executes as much as possible, and then
-    returns (a value or an error) or sleeps.
+(** Asynchronous programming with promises.
 
-    Note that inside a Lwt thread, exceptions must be raised with
-    {!fail} instead of [raise]. Also the [try ... with ...]
-    construction will not catch Lwt errors. You must use {!catch}
-    instead. You can also use {!wrap} for functions that may raise
-    normal exception.
+    {b Promises} are placeholders for values that take a long time to compute.
+    Promises are similar to [ref]s – they can store one value. Here is how they
+    differ from [ref]s:
 
-    Lwt also provides the syntax extension {!Pa_lwt} to make code
-    using Lwt more readable.
-*)
+    - A promise might not have a value yet. This is called a {e pending}
+      promise.
+    - Writing a value into a promise is called {e resolving} it.
+    - It's possible to attach {b callbacks} to a pending promise. They will be
+      run when the promise gets resolved.
+    - A promise can be resolved only once. Once a promise has a value, it's
+      immutable.
 
-(** {2 Definitions and basics} *)
+    So, promises are write-once optional references, and when they don't yet
+    have a value, they store a list of callbacks that are waiting for the value.
+
+    Why is this useful?
+
+    The waiting callbacks make promises a nice data type for asynchronous
+    programming. For example, you can ask Lwt to [read] a file. Lwt immediately
+    returns you only a {e promise} for the data.
+
+    You can neglect this promise for a while. You can do some other computation,
+    request more I/O, etc. At some point, you might decide to attach a callback
+    to the [read] promise, maybe several callbacks.
```

Comment:
I don't think this represents a typical usage of Lwt: in most cases, you attach the callbacks immediately. I.e., in typical use, there are more bind operations on unnamed promises than on named ones.

The description makes it look like you need to sort of be careful about when you bind things. Specifically, the combination of this paragraph and “In the meantime, the [read] operation is running in the background” (below) makes it sound like the concurrent operations occur because you waited to attach a callback.

Reply:
I see what you're saying, but I actually don't want to describe typical usage here. I think the typical usage obscures what Lwt actually is. It makes Lwt look like straight-line code. I think that's partially responsible for how difficult it is to "get" Lwt (even if it looks nice). My goal here, at the beginning of the docs, is to build up an accurate mental model quickly, in the simplest terms possible. Any suggestions on that are welcome.

I'll think about how to make it clear that waiting is not related to concurrency.

Reply:
And to be more specific, the "simplest possible terms" I chose were 1. promises and 2. attaching callbacks to promises, so this affects how the rest of the intro reads.

Reply:
I see what you mean by having readers build a correct mental model instead of being shown standard/usual/typical code. I still thinks that this paragraph can be misleading w.r.t. the mental model because readers may infer that binding interferes with promise resolution.

Maybe explicitly saying something along the lines of

> afterwards, two things happen (in any order):
> - you attach a callback to the promise
> - the read operation terminates which resolves the promise
>
> These operations can happen in any order. Whether you bind before or after the read terminates does not affect the program.
>
> When both of these operations have happened, the callback is called.

Reply:
That's probably a good idea. I'll probably spend a loooong time figuring out a really brief way to say that :p

Reply:
Also, I want to note, that my idea of the completed docs includes a set of examples. I hope that will be a good place to see typical usage. There should also be a tutorial, and this mental-model-building should eventually end up there, rather than in the reference docs.

Reply:
I decided to be clearer about attaching callbacks to pending promises/trying to attach them to resolved promises immediately in the bulleted list that lists differences between promises and `ref`s. Let's see how that works..